rasa_sdk/forms.py

```diff
@@ -726,57 +726,90 @@ async def extract_custom_slots(
             `SlotSet` for any extracted slots.
         """
         custom_slots = []
-        slots_to_extract = await self.required_slots(dispatcher, tracker, domain)
-
-        if slots_to_extract is None:
-            return []
+        slots_to_extract = await self.required_slots(
+            self.slots_mapped_in_domain(domain), dispatcher, tracker, domain
+        )
 
         for slot_name in slots_to_extract:
-            method_name = f"extract_{slot_name.replace('-', '_')}"
-            extract_method = getattr(self, method_name, None)
+            custom_slots += await self._extract_slot(
+                slot_name, dispatcher, tracker, domain
+            )
+        return custom_slots
 
-            if not extract_method:
-                warnings.warn(
-                    f"No method '{method_name}' found for slot "
-                    f"'{slot_name}'. Skipping extraction for this slot."
-                )
-                continue
+    async def _extract_slot(
+        self,
+        slot_name: Text,
+        dispatcher: "CollectingDispatcher",
+        tracker: "Tracker",
+        domain: "DomainDict",
+    ) -> List[EventType]:
+        method_name = f"extract_{slot_name.replace('-', '_')}"
+        also_mapped_in_domain = slot_name in self.slots_mapped_in_domain(domain)
+        extract_method = getattr(self, method_name, None)
+
+        if not extract_method and not also_mapped_in_domain:
+            warnings.warn(
+                f"No method '{method_name}' found for slot "
+                f"'{slot_name}'. Skipping extraction for this slot."
+            )
+            return []
+
+        if not extract_method:
+            return []
 
-            extracted = await utils.call_potential_coroutine(
-                extract_method(dispatcher, tracker, domain)
+        if also_mapped_in_domain:
+            warnings.warn(
+                f"Slot '{slot_name}' is mapped in the domain and your custom "
+                f"action defines '{method_name}'. It is recommended to define a "
+                f"slot mapping in either of both places."
             )
 
-            if extracted:
-                custom_slots += [
-                    SlotSet(slot, value) for slot, value in extracted.items()
-                ]
-            else:
-                warnings.warn(
-                    f"Cannot extract `{slot_name}`: make sure the extract method "
-                    f"returns the correct output."
-                )
+        extracted = await utils.call_potential_coroutine(
+            extract_method(dispatcher, tracker, domain)
+        )
 
-        return custom_slots
+        if extracted:
+            return [SlotSet(slot, value) for slot, value in extracted.items()]
+
+        warnings.warn(
+            f"Cannot extract `{slot_name}`: make sure the extract method "
+            f"returns the correct output."
+        )
```

Comment:
Hmmm, I think this warning shouldn't be here. For any user utterance, we try to extract all slots from it. So it's natural that in most cases the call to an extractor won't actually extract anything, right?

Reply:
They need to return a dict of the form `{"slotname": <value>}` so the warning makes sense. Will adapt it a bit though.

Reply:
Okay, thanks for correcting me!